clients/drcachesim/scheduler/scheduler_impl.h

```diff
@@ -570,6 +554,21 @@ template <typename RecordType, typename ReaderType> class scheduler_impl_tmpl_t
     virtual scheduler_status_t
     set_initial_schedule(std::unordered_map<int, std::vector<int>> &workload2inputs) = 0;
 
+    // When an output's input changes from a valid (not INVALID_INPUT_ORDINAL) input to
+    // something else, swap_out_input() is called on the outgoing input (whose lock is
+    // held if "caller_holds_input_lock" is true; it will never be true for
+    // MAP_TO_ANY_OUTPUT).  This is called after the input's fields (such as cur_output
+    // and last_run_time) have been updated.
+    virtual stream_status_t
+    swap_out_input(output_ordinal_t output, input_ordinal_t input, int workload,
+                   bool caller_holds_input_lock) = 0;
+
+    // When an output's input changes to a valid (not INVALID_INPUT_ORDINAL) input
+    // different from the previous input, swap_in_input() is called on the incoming
+    // input (whose lock is always held by the caller).
+    virtual stream_status_t
+    swap_in_input(output_ordinal_t output, input_ordinal_t input) = 0;
```

Comment:
Isn't `workload` useful to know here too like swap_out_input?

Reply:
The reason that was present was to avoid the callee having to acquire the lock to get the workload; the lock is already held for swap_in.  Your earlier requests added a lock to swap_out in my use case so the workload can be read there and I'm removing it from swap_out.